web/static/util.js

```diff
@@ -714,3 +714,43 @@ function _schemaResolveRefs( a_props, a_refs ){
     }
 }
 */
+
+// Note about the below, I am assuming that the client ID is readily available since it is used in datafed-ws.js
+/**
+ * Basic implementation of get_authorize_url from Globus SDK
+ *  @param {UUID} client_id The UUID of the Globus authentication client
+ *  @param {string} redirect_uri The URI safe application-wide Globus Auth redirect URI.
+ *  @param {Array<string>}requested_scopes The scopes on the token(s) being requested
+ *      In the case of accessing a mapped collection, this should include the mapped collection's UUID
+ *      like so: https://auth.globus.org/scopes/YOUR-UUID-HERE/data_access
+ *  @param {string} state Allows the application to pass information back to itself
+ *  @param {boolean} refresh_tokens Request refresh tokens in addition to access tokens
+ *  @param {object} query_params Additional params
+ *  @returns {string} The URL a user can follow to provide authorization and consent via Globus
+ */
+export function globusGetAuthorizeURL(client_id, redirect_uri, requested_scopes, state, refresh_tokens=false, query_params) {
+    // TODO: stubbed client object to provide base url, we should extract if possible
+    const auth_client = {
+        base_url: "https://auth.globus.org",
+    };
+    const authorize_base_url = auth_client.base_url + "/v2/oauth2/authorize";
+    let required_scopes = [...requested_scopes];
+    if (refresh_tokens) {
+        required_scopes = [...requested_scopes, "offline_access"];
+    }
+    let params = new URLSearchParams({
+        "client_id": client_id,
+        "redirect_uri": redirect_uri,
+        "scope": required_scopes.join(" "),    // Scopes need to be separated by a space
+        "state": state,
+        "response_type": "code",
+        "access_type": refresh_tokens ? "offline" : "online",
+        "prompt": "login",
+    });
+    Object.entries(query_params).forEach(([key, value]) => {
+       if (!!value && !params.get(key)) { // short-circuit on empty param values or if param already defined, TODO: are there cases where we may want empty params?
+           params.set(key, value);
+       }

```

Comment:
I cannot think of an instance where you would not be providing a client_id.

Reply:
This bit is iterating over the additional `query_params` which are currently not being allowed to overwrite the function's parameters. I think, to your point, ensuring some parameters were passed and not-null, such as `client_id` or `redirect_uri` could have value. State is totally optional. I should default it like the reference code, however.